src/index.ts

```diff
@@ -121,12 +125,29 @@ export class Pipeline implements Serializable {
       : {
           explicitDependencies: false,
         };
-
+    const steps = await this.toList(newOpts);
+    if (opts.mutator) {
+      for (const step of steps) {
+        if (step instanceof Step) {
+          const deps = {
+            dependencies: cloneDeep(step.dependencies),
+            effectDependencies: cloneDeep(step.effectDependencies),
```

Comment:
Are the objects in these arrays not of type `Step`? I don't think we should clone this deeply, it means the contents of each array are cloned as well.

Idea: Can we `Object.freeze` the dependencies and `effectDependencies`?
Or alternatively we can just assert that the contents of each of these arrays are the same before and after the mutation. That means we don't need to clone deeply:

```suggestion
            dependencies: [...step.dependencies],
            effectDependencies: [...step.effectDependencies],
```
which allows a mutator to read and tamper with each of the contents of these arrays, but maybe that's okay, they would have access to them anyway.

Reply:
I ended up using `new Set(dep)` to do this instead, I think this accomplishes the same thing https://github.com/joscha/buildkite-graph/pull/75/commits/24a1f0375bde5eaaf9bc8a95d3760f5c413dfe78#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L133

Reply:
It's not the same as `[...array]` because it doesn't preserve the order (though that isn't significant in this case) and `isEqual` doesn't have a special code path for sets, but it does for arrays.

I'd rather use `[...array]` unless there's a specific reason to prefer sets, like we purposefully want to ignore changing the order of the elements for example.

Reply:
I thought sets were unordered,  but i see that isn't necessarily true. Will change this to your suggestion

Reply:
Ohhh no wait, I see, I thought dependencies and effects were arrays, but they're actually sets, so you were right, you can leave it as `new Set(...)` for me.